File: helpers/pyband/tests/transaction_test.py
import requests
from pyband.client import Client
import pytest

from pyband import Transaction
from pyband.message import MsgRequest
from pyband.wallet import Address, PrivateKey
from pyband.client import Client
from pyband.exceptions import EmptyMsgError, UndefinedError, ValueTooLargeError

TEST_RPC = "https://api-mock.bandprotocol.com/rest/"

client = Client(TEST_RPC)


def test_get_sign_data_success():
    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_account_num(100)
        .with_sequence(30)
        .with_chain_id("bandchain")
    )

    assert (
        t.get_sign_data()
        == b'{"account_number":"100","chain_id":"bandchain","fee":{"amount":[{"amount":"0","denom":"uband"}],"gas":"200000"},"memo":"","msgs":[{"type":"oracle/Request","value":{"ask_count":"4","calldata":"AAAAA0JUQwAAAAAAAAAB","client_id":"from_pyband","min_count":"3","oracle_script_id":"1","sender":"band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"}}],"sequence":"30"}'
    )


def test_get_sign_data_with_gas_success():
    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_account_num(100)
        .with_sequence(30)
        .with_chain_id("bandchain")
        .with_gas(500000)
        .with_fee(10)
    )

    assert (
        t.get_sign_data()
        == b'{"account_number":"100","chain_id":"bandchain","fee":{"amount":[{"amount":"10","denom":"uband"}],"gas":"500000"},"memo":"","msgs":[{"type":"oracle/Request","value":{"ask_count":"4","calldata":"AAAAA0JUQwAAAAAAAAAB","client_id":"from_pyband","min_count":"3","oracle_script_id":"1","sender":"band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"}}],"sequence":"30"}'
    )


def test_get_sign_data_with_auto_success(requests_mock):

    requests_mock.register_uri(
        "GET",
        "{}/auth/accounts/band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c".format(TEST_RPC),
        json={
            "height": "650788",
            "result": {
                "type": "cosmos-sdk/Account",
                "value": {
                    "address": "band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c",
                    "coins": [{"denom": "uband", "amount": "104082359107"}],
                    "public_key": {
                        "type": "tendermint/PubKeySecp256k1",
                        "value": "A/5wi9pmUk/SxrzpBoLjhVWoUeA9Ku5PYpsF3pD1Htm8",
                    },
                    "account_number": "36",
                    "sequence": "927",
                },
            },
        },
        status_code=200,
    )

    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_auto(client)
        .with_chain_id("bandchain")
        .with_gas(500000)
        .with_fee(10)
    )

    assert (
        t.get_sign_data()
        == b'{"account_number":"36","chain_id":"bandchain","fee":{"amount":[{"amount":"10","denom":"uband"}],"gas":"500000"},"memo":"","msgs":[{"type":"oracle/Request","value":{"ask_count":"4","calldata":"AAAAA0JUQwAAAAAAAAAB","client_id":"from_pyband","min_count":"3","oracle_script_id":"1","sender":"band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"}}],"sequence":"927"}'
    )


def test_create_transaction_with_auto_fail(requests_mock):

    requests_mock.register_uri(
        "GET",
        "{}/auth/accounts/band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c".format(TEST_RPC),
        json={
            "height": "650788",
            "result": {
                "type": "cosmos-sdk/Account",
                "value": {
                    "address": "band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c",
                    "coins": [{"denom": "uband", "amount": "104082359107"}],
                    "public_key": {
                        "type": "tendermint/PubKeySecp256k1",
                        "value": "A/5wi9pmUk/SxrzpBoLjhVWoUeA9Ku5PYpsF3pD1Htm8",
                    },
                    "account_number": "36",
                    "sequence": "927",
                },
            },
        },
        status_code=200,
    )

    with pytest.raises(EmptyMsgError, match="messsage is empty, please use with_messages at least 1 message"):
        t = Transaction().with_auto(client)


def test_get_sign_data_with_memo_success():
    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_account_num(100)
        .with_sequence(30)
        .with_chain_id("bandchain")
        .with_memo("from_pyband_test")
    )

    assert (
        t.get_sign_data()
        == b'{"account_number":"100","chain_id":"bandchain","fee":{"amount":[{"amount":"0","denom":"uband"}],"gas":"200000"},"memo":"from_pyband_test","msgs":[{"type":"oracle/Request","value":{"ask_count":"4","calldata":"AAAAA0JUQwAAAAAAAAAB","client_id":"from_pyband","min_count":"3","oracle_script_id":"1","sender":"band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"}}],"sequence":"30"}'
    )


def test_get_sign_data_msgs_fail():
    t = Transaction().with_account_num(100).with_sequence(30).with_chain_id("bandchain")

    with pytest.raises(EmptyMsgError, match="message is empty"):
        t.get_sign_data()


def test_get_sign_data_account_num_fail():
    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_sequence(30)
        .with_chain_id("bandchain")
    )

    with pytest.raises(UndefinedError, match="account_num should be defined"):
        t.get_sign_data()


def test_get_sign_data_sequence_fail():
    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_account_num(100)
        .with_chain_id("bandchain")
    )

    with pytest.raises(UndefinedError, match="sequence should be defined"):
        t.get_sign_data()


def test_get_sign_data_chain_id_fail():
    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_account_num(100)
        .with_sequence(30)
    )

    with pytest.raises(UndefinedError, match="chain_id should be defined"):
        t.get_sign_data()


def test_get_sign_data_memo_fail():
    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=Address.from_acc_bech32("band13eznuehmqzd3r84fkxu8wklxl22r2qfmtlth8c"),
            )
        )
        .with_account_num(100)
        .with_sequence(30)
        .with_chain_id("bandchain")
    )

    with pytest.raises(ValueTooLargeError, match="memo is too large"):
        t.with_memo(
            "This is the longest memo in the world. This is the longest memo in the world. This is the longest memo in the world. This is the longest memo in the world. This is the longest memo in the world. This is the longest memo in the world. This is the longest memo in the world.This is the longest memo in the world. This is the longest memo in the world.This is the longest memo in the world."
        )


def test_get_tx_data_success():
    priv = PrivateKey.from_mnemonic("s")
    pubkey = priv.to_pubkey()
    addr = pubkey.to_address()

    t = (
        Transaction()
        .with_messages(
            MsgRequest(
                oracle_script_id=1,
                calldata=bytes.fromhex("000000034254430000000000000001"),
                ask_count=4,
                min_count=3,
                client_id="from_pyband",
                sender=addr,
            )
        )
        .with_account_num(100)
        .with_sequence(30)
        .with_chain_id("bandchain")
    )

    raw_data = t.get_sign_data()
    signature = priv.sign(raw_data)
    raw_tx = t.get_tx_data(signature, pubkey)

    assert raw_tx == {
        "msg": [
            {
                "type": "oracle/Request",
                "value": {
                    "oracle_script_id": "1",
                    "calldata": "AAAAA0JUQwAAAAAAAAAB",
                    "ask_count": "4",
                    "min_count": "3",
                    "client_id": "from_pyband",
                    "sender": "band1jrhuqrymzt4mnvgw8cvy3s9zhx3jj0dq30qpte",
                },
            }
        ],
        "fee": {"gas": "200000", "amount": [{"denom": "uband", "amount": "0"}]},
        "memo": "",
        "signatures": [
            {
                "signature": "aFvTgkY9F14dHfp2mtq8V2eCTKUtY1T9WKs99jmR8To3JB16cawbmpa1TRUdzfnqLXBh+o6XUuF4bHWR6xbCCw==",
                "pub_key": {
                    "type": "tendermint/PubKeySecp256k1",
                    "value": "A/5wi9pmUk/SxrzpBoLjhVWoUeA9Ku5PYpsF3pD1Htm8",
                },
                "account_number": "100",
                "sequence": "30",
            }
        ],
    }

File: helpers/bandchain2.js/__tests__/transaction.test.ts
import { Message, Data, Transaction, Wallet } from '../src/index'
import { Address } from '../src/wallet'

const { MsgSend } = Message
const { Coin } = Data
const { PrivateKey } = Wallet

describe('Transaction', () => {
  const coin = new Coin(100000, 'uband')
  const msgSend = new MsgSend(
    Address.fromAccBech32('band1jrhuqrymzt4mnvgw8cvy3s9zhx3jj0dq30qpte'),
    Address.fromAccBech32('band1ksnd0f3xjclvg0d4z9w0v9ydyzhzfhuy47yx79'),
    [coin],
  )

  it('create successfully', () => {
    const tsc = new Transaction()

    expect(tsc.accountNum).toBeUndefined()
    expect(tsc.chainID).toBeUndefined()
    expect(tsc.sequence).toBeUndefined()
    expect(tsc.fee).toEqual(0)
    expect(tsc.gas).toEqual(200000)
    expect(tsc.withMessages.length).toEqual(0)
  })

  it('able to work "with" function', () => {
    const tsc = new Transaction()
      .withMessages(msgSend)
      .withAccountNum(100)
      .withSequence(30)
      .withChainID('bandchain')
      .withGas(500000)
      .withFee(10)
      .withMemo('bandchain2.js test')

    expect(tsc.msgs.length).toEqual(1)
    expect(tsc.accountNum).toEqual(100)
    expect(tsc.sequence).toEqual(30)
    expect(tsc.chainID).toEqual('bandchain')
    expect(tsc.fee).toEqual(10)
    expect(tsc.gas).toEqual(500000)
    expect(tsc.memo).toEqual('bandchain2.js test')
  })

  it('get error from checking integer', () => {
    const tsc = new Transaction().withMessages(msgSend)

    expect(() => {
      tsc.withAccountNum(100.5)
    }).toThrowError('accountNum is not an integer')

    expect(() => {
      tsc.withSequence(100.5)
    }).toThrowError('sequence is not an integer')

    expect(() => {
      tsc.withFee(100.5)
    }).toThrowError('fee is not an integer')

    expect(() => {
      tsc.withGas(100.5)
    }).toThrowError('gas is not an integer')
  })

  it('getSignData successfully', () => {
    const tsc = new Transaction()
      .withMessages(msgSend)
      .withAccountNum(100)
      .withSequence(30)
      .withChainID('bandchain')
      .withGas(500000)
      .withFee(10)
      .withMemo('bandchain2.js test')

    let expectedResult =
      '7b226163636f756e745f6e756d626572223a22313030222c22636861696e5f6964223a2262616e64636861696e222c22666565223a7b22616d6f756e74223a5b7b22616d6f756e74223a223130222c2264656e6f6d223a227562616e64227d5d2c22676173223a22353030303030227d2c226d656d6f223a2262616e64636861696e322e6a732074657374222c226d736773223a5b7b2274797065223a22636f736d6f732d73646b2f4d736753656e64222c2276616c7565223a7b22616d6f756e74223a5b7b22616d6f756e74223a22313030303030222c2264656e6f6d223a227562616e64227d5d2c2266726f6d5f61646472657373223a2262616e64316a7268757172796d7a74346d6e766777386376793373397a6878336a6a306471333071707465222c22746f5f61646472657373223a2262616e64316b736e64306633786a636c76673064347a39773076397964797a687a66687579343779783739227d7d5d2c2273657175656e6365223a223330227d'

    expect(tsc.getSignData().toString('hex')).toEqual(expectedResult)
  })

  it('getSignData fail', () => {
    let tsc = new Transaction()

    expect(() => tsc.getSignData()).toThrowError('message is empty')

    tsc = tsc.withMessages(msgSend)
    expect(() => tsc.getSignData()).toThrowError('accountNum should be defined')

    tsc = tsc.withAccountNum(100)
    expect(() => tsc.getSignData()).toThrowError('sequence should be defined')

    tsc = tsc.withSequence(30)
    expect(() => tsc.getSignData()).toThrowError('chainID should be defined')

    tsc = tsc.withChainID('bandchain')
    let expectedResult =
      '7b226163636f756e745f6e756d626572223a22313030222c22636861696e5f6964223a2262616e64636861696e222c22666565223a7b22616d6f756e74223a5b7b22616d6f756e74223a2230222c2264656e6f6d223a227562616e64227d5d2c22676173223a22323030303030227d2c226d656d6f223a22222c226d736773223a5b7b2274797065223a22636f736d6f732d73646b2f4d736753656e64222c2276616c7565223a7b22616d6f756e74223a5b7b22616d6f756e74223a22313030303030222c2264656e6f6d223a227562616e64227d5d2c2266726f6d5f61646472657373223a2262616e64316a7268757172796d7a74346d6e766777386376793373397a6878336a6a306471333071707465222c22746f5f61646472657373223a2262616e64316b736e64306633786a636c76673064347a39773076397964797a687a66687579343779783739227d7d5d2c2273657175656e6365223a223330227d'
    expect(tsc.getSignData().toString('hex')).toEqual(expectedResult)
  })

  it('getTxData successfully', () => {
    const tsc = new Transaction()
      .withMessages(msgSend)
      .withAccountNum(100)
      .withSequence(30)
      .withChainID('bandchain')
      .withGas(200000)

    const rawData = tsc.getSignData()
    const privKey = PrivateKey.fromMnemonic('s')
    const pubkey = privKey.toPubkey()
    const signature = privKey.sign(rawData)
    const rawTx = tsc.getTxData(signature, pubkey)

    expect(rawTx).toEqual({
      msg: [
        {
          type: 'cosmos-sdk/MsgSend',
          value: {
            amount: [
              {
                amount: '100000',
                denom: 'uband',
              },
            ],
            from_address: 'band1jrhuqrymzt4mnvgw8cvy3s9zhx3jj0dq30qpte',
            to_address: 'band1ksnd0f3xjclvg0d4z9w0v9ydyzhzfhuy47yx79',
          },
        },
      ],
      fee: { gas: '200000', amount: [{ denom: 'uband', amount: '0' }] },
      memo: '',
      signatures: [
        {
          signature:
            'LLnpUzoB7gmQd+9XxQiRyiEv3O04FWWgyX2Agm2xOoMF0iwab3BzG1L5Szl0OGfJezmMm016gF/Gjy0l9niB5w==',
          pub_key: {
            type: 'tendermint/PubKeySecp256k1',
            value: 'A/5wi9pmUk/SxrzpBoLjhVWoUeA9Ku5PYpsF3pD1Htm8',
          },
          account_number: '100',
          sequence: '30',
        },
      ],
    })
  })
})
